rtrlib/rtr_mgr.c

```diff
@@ -331,35 +379,41 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
 		goto err;
 	spki_table_init(spki_table, spki_update_fp);
 
-	for (unsigned int i = 0; i < config->len; i++) {
-		struct rtr_mgr_group cg = config->groups[i];
+	config->pfx_table = pfxt;
+	config->spki_table = spki_table;
 
-		cg.status = RTR_MGR_CLOSED;
-		if ((i > 0) && (cg.preference == last_preference)) {
-			MGR_DBG1("Error Same preference for 2 socket groups!");
-			goto err;
-		}
-		if (cg.sockets_len == 0) {
-			MGR_DBG1("Error Empty sockets array in socket group!");
-			goto err;
-		}
-		for (unsigned int j = 0; j < cg.sockets_len; j++) {
-			if (rtr_init(cg.sockets[j], NULL,
+	/* Copy the groups from the array into linked list config->groups */
+	config->len = groups_len;
+	config->groups = NULL;
+	struct rtr_mgr_group *cg;
+	struct rtr_mgr_group_node *group_node;
+
+	for (unsigned int i = 0; i < groups_len; i++) {
+		cg = malloc(sizeof(struct rtr_mgr_group));
+		memcpy(cg, &groups[i], sizeof(struct rtr_mgr_group));
+
+		cg->status = RTR_MGR_CLOSED;
+		for (unsigned int j = 0; j < cg->sockets_len; j++) {
+			if (rtr_init(cg->sockets[j], NULL,
 				     pfxt, spki_table, refresh_interval,
 				     expire_interval, retry_interval,
-				     rtr_mgr_cb, config) != RTR_SUCCESS) {
+				     rtr_mgr_cb, config, cg) != RTR_SUCCESS) {
 				err_code = RTR_INVALID_PARAM;
```

Comment:
instead of assigning err_code to the (currently) only error code that rtr_init() can return it would be better to do something like:
```
err = rtr_init();
if (err)
    goto [..]
```
If rtr_init() would change in the future to also return other errors, it's easy to miss to update this place here to ensure that the user gets the right error code back :-)